Client cannot connect to cluster on a specific port

Hello, I am trying to create a Hazelcast cluster and I ran into a problem.
Trying to generate a member cluster in a specific port (5710 i.e.), I connect a client to this cluster on this hardcoded port (:5710) into the method addAddress, but when I execute the client it fails trying to access the cluster. 
If I don't specify port (only localhost is set at client configuration) or I set listening port to 5701 for cluster and client (not any port else) the client does its job succesfully. I wonder if there is any kind of bug or could be possible a change from the version of Hazelcast. 
call is like this one:
```
Member:
	ClientConfig clientConfig = new ClientConfig();
	clientConfig.getNetworkConfig().setPublicAddress("127.0.0.1").setPort(5710);
Client:
    	ClientConfig clientConfig = new ClientConfig();
	clientConfig.getNetworkConfig().addAddress("127.0.0.1:5710");
	HazelcastInstance hz = HazelcastClient.newHazelcastClient(clientConfig);
```
With 3.1 version setPort is working fine, but it doesnt at new versions: config.getNetworkConfig().setPublicAddress("127.0.0.1").setPort(5710);

I tried also an example (the multimap example from the hazelcast-code-samples repo) with versions 3.4.1 and 3.5. I get this error on both:

------------------
Member trace:
------------------
```
Members [1] {
	Member [127.0.0.1]:0 this
}

feb 24, 2015 5:37:20 PM com.hazelcast.core.LifecycleService
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Address[127.0.0.1]:0 is STARTED
feb 24, 2015 5:37:20 PM com.hazelcast.partition.InternalPartitionService
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Initializing cluster partition table first arrangement...
PutMember:Done
feb 24, 2015 5:37:43 PM com.hazelcast.nio.tcp.SocketAcceptor
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Accepting socket connection from /127.0.0.1:54249
feb 24, 2015 5:37:43 PM com.hazelcast.nio.tcp.TcpIpConnectionManager
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Established socket connection between /127.0.0.1:5710 and 127.0.0.1/127.0.0.1:54249
feb 24, 2015 5:37:43 PM com.hazelcast.client.impl.client.AuthenticationRequest
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Received auth from Connection [/127.0.0.1:5710 -> 127.0.0.1/127.0.0.1:54249], endpoint=null, live=true, type=JAVA_CLIENT, successfully authenticated
feb 24, 2015 5:38:24 PM com.hazelcast.nio.tcp.TcpIpConnection
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Connection [Address[127.0.0.1]:54249] lost. Reason: java.io.IOException[Se ha forzado la interrupción de una conexión existente por el host remoto]
feb 24, 2015 5:38:24 PM com.hazelcast.nio.tcp.ReadHandler
ADVERTENCIA: [127.0.0.1]:0 [dev] [3.4.1] hz._hzInstance_1_dev.IO.thread-in-0 Closing socket to endpoint Address[127.0.0.1]:54249, Cause:java.io.IOException: Se ha forzado la interrupción de una conexión existente por el host remoto
feb 24, 2015 5:38:24 PM com.hazelcast.client.ClientEndpointManager
INFORMACIÓN: [127.0.0.1]:0 [dev] [3.4.1] Destroying ClientEndpoint{conn=Connection [0.0.0.0/0.0.0.0:5710 -> 127.0.0.1/127.0.0.1:54249], endpoint=Address[127.0.0.1]:54249, live=false, type=JAVA_CLIENT, principal='ClientPrincipal{uuid='31b415cd-e8a5-4005-94a7-133b0018f31d', ownerUuid='e8a04fb7-7517-4902-b268-75e3bf22c643'}', firstConnection=true, authenticated=true}

----------------
Client trace:
----------------
Members [1] {
	Member [127.0.0.1]:0
}

feb 24, 2015 5:43:43 PM com.hazelcast.client.spi.ClientPartitionService
GRAVE: Error while fetching cluster partition table!
java.net.BindException: Cannot assign requested address: connect
	at sun.nio.ch.Net.connect0(Native Method)
	at sun.nio.ch.Net.connect(Net.java:457)
	at sun.nio.ch.Net.connect(Net.java:449)
	at sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:647)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:102)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:386)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:353)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)

Exception in thread "main" java.lang.IllegalStateException: Cannot get initial partitions!
	at com.hazelcast.client.spi.impl.ClientPartitionServiceImpl.getInitialPartitions(ClientPartitionServiceImpl.java:89)
	at com.hazelcast.client.spi.impl.ClientPartitionServiceImpl.start(ClientPartitionServiceImpl.java:65)
	at com.hazelcast.client.impl.HazelcastClientInstanceImpl.start(HazelcastClientInstanceImpl.java:189)
	at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:81)
	at PrintMember.main(PrintMember.java:16)
```
If I substitute this line in the Member
config.getNetworkConfig().setPublicAddress("127.0.0.1").setPort(5710);

for this:
config.getNetworkConfig().setPublicAddress("127.0.0.1:5710");

The trace I get is this for the client (the client cant reach the cluster):

---------------
Client trace:
---------------
````
feb 24, 2015 5:46:30 PM com.hazelcast.core.LifecycleService
INFORMACIÓN: HazelcastClient[hz.client_0_dev][3.4.1] is STARTING
feb 24, 2015 5:46:30 PM com.hazelcast.core.LifecycleService
INFORMACIÓN: HazelcastClient[hz.client_0_dev][3.4.1] is STARTED
feb 24, 2015 5:46:31 PM com.hazelcast.client.spi.impl.ClusterListenerThread
ADVERTENCIA: Unable to get alive cluster connection, try in 1987 ms later, attempt 1 of 2.
feb 24, 2015 5:46:34 PM com.hazelcast.client.spi.impl.ClusterListenerThread
ADVERTENCIA: Unable to get alive cluster connection, try in 1999 ms later, attempt 2 of 2.
feb 24, 2015 5:46:36 PM com.hazelcast.client.spi.impl.ClusterListenerThread
GRAVE: Error while connecting to cluster!
java.lang.IllegalStateException: Unable to connect to any address in the config! The following addresses were tried:[/127.0.0.1:5710]
	at com.hazelcast.client.spi.impl.ClusterListenerThread.connectToOne(ClusterListenerThread.java:286)
	at com.hazelcast.client.spi.impl.ClusterListenerThread.run(ClusterListenerThread.java:89)
Caused by: java.net.ConnectException: Connection refused: no further information
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:386)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:353)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)

feb 24, 2015 5:46:36 PM com.hazelcast.core.LifecycleService
INFORMACIÓN: HazelcastClient[hz.client_0_dev][3.4.1] is SHUTTING_DOWN
feb 24, 2015 5:46:36 PM com.hazelcast.core.LifecycleService
INFORMACIÓN: HazelcastClient[hz.client_0_dev][3.4.1] is SHUTDOWN
Exception in thread "main" java.lang.IllegalStateException: Unable to connect to any address in the config! The following addresses were tried:[/127.0.0.1:5710]
	at com.hazelcast.client.spi.impl.ClusterListenerThread.connectToOne(ClusterListenerThread.java:286)
	at com.hazelcast.client.spi.impl.ClusterListenerThread.run(ClusterListenerThread.java:89)
Caused by: java.net.ConnectException: Connection refused: no further information
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at sun.nio.ch.SocketAdaptor.connect(SocketAdaptor.java:111)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:386)
	at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$ConnectionProcessor.call(ClientConnectionManagerImpl.java:353)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
```

Thanks in advance.

--- a/hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java
+++ b/hazelcast/src/main/java/com/hazelcast/instance/DefaultAddressPicker.java
@@ -80,6 +80,9 @@ class DefaultAddressPicker implements AddressPicker {
                 logger.finest("Using public address the same as the bind address: " + publicAddress);
             }
         } catch (Exception e) {
+            if (serverSocketChannel != null) {
+                serverSocketChannel.close();
+            }
             logger.severe(e);
             throw e;
         }
@@ -296,7 +299,7 @@ class DefaultAddressPicker implements AddressPicker {
         if (address != null) {
             address = address.trim();
             if ("127.0.0.1".equals(address) || "localhost".equals(address)) {
-                return pickLoopbackAddress();
+                return pickLoopbackAddress(defaultPort);
             } else {
                 // allow port to be defined in same string in the form of <host>:<port>, e.g. 10.0.0.0:1234
                 AddressUtil.AddressHolder holder = AddressUtil.getAddressHolder(address, defaultPort);
@@ -308,6 +311,11 @@ class DefaultAddressPicker implements AddressPicker {
 
     private AddressDefinition pickLoopbackAddress() throws UnknownHostException {
         return new AddressDefinition(InetAddress.getByName("127.0.0.1"));
+    }
+
+    private AddressDefinition pickLoopbackAddress(int defaultPort) throws UnknownHostException {
+        InetAddress adddress = InetAddress.getByName("127.0.0.1");
+        return new AddressDefinition(adddress.getHostAddress(), defaultPort, adddress);
     }
 
     private AddressDefinition pickMatchingAddress(Collection<InterfaceDefinition> interfaces) throws SocketException {

--- a/hazelcast/src/test/java/com/hazelcast/cluster/PublicAddressTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/PublicAddressTest.java
@@ -1,0 +1,107 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cluster;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.nio.Address;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.net.UnknownHostException;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class})
+public class PublicAddressTest {
+
+    public static final int DEFAULT_PORT = 5701;
+
+    @After
+    public void cleanup() {
+        Hazelcast.shutdownAll();
+    }
+
+    @Test
+    public void testUseDefaultPortWhenLoopback() throws UnknownHostException {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("127.0.0.1");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("127.0.0.1", DEFAULT_PORT), instance.getCluster().getLocalMember().getAddress());
+    }
+
+    @Test
+    public void testUseDefaultPortWhenLocalhost() throws UnknownHostException {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("localhost");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("127.0.0.1", DEFAULT_PORT), instance.getCluster().getLocalMember().getAddress());
+    }
+
+    @Test
+    public void testUseSpecifiedHost() throws UnknownHostException {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("www.example.org");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("www.example.org", DEFAULT_PORT), instance.getCluster().getLocalMember().getAddress());
+    }
+
+    @Test
+    public void testUseSpecifiedHostAndPort() throws UnknownHostException {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("www.example.org:6789");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("www.example.org", 6789), instance.getCluster().getLocalMember().getAddress());
+    }
+
+    @Test
+    public void testUseSpecifiedHostAndPortViaProperty() throws UnknownHostException {
+        Config config = new Config();
+        config.setProperty("hazelcast.local.publicAddress", "192.168.1.1:6789");
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("192.168.1.1", 6789), instance.getCluster().getLocalMember().getAddress());
+    }
+
+    @Test(expected = HazelcastException.class)
+    public void testInvalidPublicAddress() {
+        Config config = new Config();
+        config.getNetworkConfig().setPublicAddress("invalid");
+
+        Hazelcast.newHazelcastInstance(config);
+    }
+
+    @Test
+    public void testUseBindAddressWhenNoPublicAddressGiven() throws UnknownHostException {
+        Config config = new Config();
+        config.getNetworkConfig().setPort(5705);
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+
+        assertEquals(new Address("127.0.0.1", 5705), instance.getCluster().getLocalMember().getAddress());
+    }
+}

